### PR TITLE
Instanced Utils

### PR DESCRIPTION
**Pull request by AzureAaron**
Adds some utility methods that "auto get" `equals`, `hashCode` and `toString` implementations. Just thought this could be useful with the growing number of instanced classes we're using that aren't records.

### CODE REVIEW

File: src/main/java/de/hysky/skyblocker/utils/InstancedUtils.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+
+/**
+ * @implNote If implementing any of these onto a class, ensure that all subclasses have an implementation of the methods too.
+ */
+public class InstancedUtils {
+	private static final Logger LOGGER = LogUtils.getLogger();
+	private static final Map<Class<?>, MethodHandle> EQUALS_CACHE = new ConcurrentHashMap<>();
+	private static final Map<Class<?>, MethodHandle> HASH_CODE_CACHE = new ConcurrentHashMap<>();
+	private static final Map<Class<?>, MethodHandle> TO_STRING_CACHE = new ConcurrentHashMap<>();
```
Comment (AzureAaron) [inline src/main/java/de/hysky/skyblocker/utils/InstancedUtils.java:25]: Depending on what we want, caching could instead be done by the caller class in constants (which may actually be faster).

File: src/test/java/de/hysky/skyblocker/utils/InstancedUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+			this.y = y;
+			this.z = z;
+		}
+		
+		@Override
+		public boolean equals(Object o) {
+			try {
+				return (boolean) InstancedUtils.equals(getClass()).invokeExact(this, o);
+			} catch (Throwable ignored) {
+				return false;
```
Comment (kevinthegreat1) [inline src/test/java/de/hysky/skyblocker/utils/InstancedUtilsTest.java:71]: Maybe we should throw or return default equals?

**Review by kevinthegreat1 — APPROVED** [review]
Cursed but sure ig... Also, should we throw instead of returning the default value?

File: src/test/java/de/hysky/skyblocker/utils/InstancedUtilsTest.java (hunk 2 above)
Comment (AzureAaron) [reply src/test/java/de/hysky/skyblocker/utils/InstancedUtilsTest.java:71]: Perhaps the default equals - either way this should never throw an exception just that the compiler doesn’t allow you to ignore it.

**Comment by AzureAaron** [comment]
> Cursed but sure ig... Also, should we throw instead of returning the default value?

Kinda cursed yes but I think the most useful thing here is probably the hash code thing since we seem to use a lot of HashMaps or things that depend on a unique hash code.

**Comment by AzureAaron** [comment]
After looking into it I don't think we can throw from those methods unless we do a runtime exception or some other unchecked one, would that be desirable?

**Review by kevinthegreat1 — APPROVED** [review]

**Pull request merged**